pkg/subscriptions/subscriptions-client.js

```diff
@@ -0,0 +1,416 @@
+                  '--proxypass', subscriptionDetails.proxyPass);
+    }
+
+    // only pass org info if user provided it
+    if (subscriptionDetails.org)
+        args.push('--org', subscriptionDetails.org);
+
+    /* TODO DBus API doesn't deliver what we need, so we call subscription manager
+     * without translations and parse the output
+     */
+    var process = cockpit.spawn(args, {
+        directory: '/',
+        superuser: "require",
+        environ: ['LC_ALL=C'],
+        err: "out"
+    });
+
+    // once we start this, we're in an unknown state
+    client.subscriptionStatus.products = [];
+    client.subscriptionStatus.status = undefined;
+
+    var promise;
+    var buffer = '';
+    process
+        .input('')
+        .stream(function(text) {
+            buffer += text;
+        })
+        .done(function(output) {
+            deferred.resolve();
+        })
+        .fail(function(ex) {
+            if (ex.problem === "cancelled") {
+                deferred.reject(ex);
+                return;
+            }
+
+            /* detect error types we recognize, fall back is generic error */
+            var invalidUsernameString = 'Invalid username or password.';
+            var invalidCredentialsString = 'Invalid Credentials';
+            var message = buffer.trim();
+            if (message.indexOf(invalidUsernameString) !== -1) {
+                message = cockpit.format("$0 ($1)", _("Invalid username or password"), message.substring(invalidUsernameString.length).trim());
+            } else if (message.indexOf(invalidCredentialsString) !== -1) {
+                message = cockpit.format("$0 ($1)", _("Invalid credentials"), message.substring(invalidCredentialsString.length).trim());
+            } else if ((message.indexOf('EOF') === 0) && (message.indexOf('Organization:') !== -1)) {
+                message = _("'Organization' required to register.");
+            } else if ((message.indexOf('EOF') === 0) && (message.indexOf('Username:') !== -1)) {
+                message = _("Login/password or activation key required to register.");
+            } else if (message.indexOf('Must provide --org with activation keys') !== -1) {
+                message = _("'Organization' required when using activation keys.");
+            } else if (message.indexOf('The system has been registered') !== -1) {
+                /*
+                 * Currently we don't separate registration & subscription.
+                 * Our auto-attach may have failed, so close the dialog and
+                 * update status.
+                 */
+                deferred.resolve();
+                return;
+            }
+            requestUpdate();
+            var error = new Error(message);
+            deferred.reject(error);
+        });
+
+    promise = deferred.promise();
+    promise.cancel = function cancel() {
+        process.close("cancelled");
+        // we have no idea what the current state is
+        requestUpdate();
```

Comment:
> Oh this is probably bound to the dialog stuff ... just a guess. In which case it makes sense.

Cancel is triggered by the user in the dialog. At this point, there is no telling how far the register process got, so we need an update to make sure that we have a consistent state.